plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataBuilder.java

```diff
@@ -0,0 +1,356 @@
+          coordinates.getVersion()
+      );
+      snapshots.add(snapshotVersion);
+    }
+    return Maven2Metadata.newVersionLevel(
+        DateTime.now(),
+        getGroupId(),
+        getArtifactId(),
+        getBaseVersion(),
+        latestVersionCoordinates.coordinates.getTimestamp(),
+        latestVersionCoordinates.coordinates.getBuildNumber(),
+        snapshots
+    );
+  }
+
+  public void addArtifactVersion(final MavenPath mavenPath) {
+    checkNotNull(mavenPath, "mavenPath");
+    if (mavenPath.isSubordinate() || mavenPath.getCoordinates() == null) {
+      return;
+    }
+    checkState(Objects.equals(getGroupId(), mavenPath.getCoordinates().getGroupId()));
+    checkState(Objects.equals(getArtifactId(), mavenPath.getCoordinates().getArtifactId()));
+    checkState(Objects.equals(getBaseVersion(), mavenPath.getCoordinates().getBaseVersion()));
+
+    log.debug("Discovered {}:{}:{}:{}:{}",
+        mavenPath.getCoordinates().getGroupId(),
+        mavenPath.getCoordinates().getArtifactId(),
+        mavenPath.getCoordinates().getVersion(),
+        mavenPath.getCoordinates().getClassifier(),
+        mavenPath.getCoordinates().getExtension());
+
+    addBaseVersion(mavenPath.getCoordinates().getBaseVersion());
+
+    if (!mavenPath.getCoordinates().isSnapshot()) {
+      return;
+    }
+    if (Objects.equals(mavenPath.getCoordinates().getBaseVersion(), mavenPath.getCoordinates().getVersion())) {
+      log.warn("Non-timestamped snapshot, ignoring it: {}", mavenPath);
+      return;
+    }
+
+    final String key = key(mavenPath.getCoordinates());
+    final Version version = parseVersion(mavenPath.getCoordinates().getVersion());
+    final VersionCoordinates versionCoordinates = new VersionCoordinates(version, mavenPath.getCoordinates());
+
+    // maintain latestVersionCoordinates
+    if (latestVersionCoordinates == null || latestVersionCoordinates.version.compareTo(version) < 0) {
+      latestVersionCoordinates = versionCoordinates;
+    }
+
+    // maintain latestVersionCoordinatesMap
+    if (!latestVersionCoordinatesMap.containsKey(key)) {
+      latestVersionCoordinatesMap.put(key, versionCoordinates);
+    }
+    else {
+      // add if contained version is less than version
+      final VersionCoordinates other = latestVersionCoordinatesMap.get(key);
+      if (other.version.compareTo(versionCoordinates.version) < 0) {
+        latestVersionCoordinatesMap.put(key, versionCoordinates);
+      }
+    }
+  }
+
+  private String key(final Coordinates coordinates) {
+    if (coordinates.getClassifier() == null) {
+      return coordinates.getExtension();
+    }
+    else {
+      return coordinates.getExtension() + ":" + coordinates.getClassifier();
+    }
+  }
+
+  private Version parseVersion(final String version) {
+    try {
+      return versionScheme.parseVersion(version);
+    }
+    catch (InvalidVersionSpecificationException e) {
+      // nada
+      throw Throwables.propagate(e); // make IDEA happy wrt NPE
```

Comment:
log the parse error?  also not sure about adding code just to make IDEs happy - presumably it's seen something in its static code-analysis which might deserve a closer look :)